### DISCUSSION TITLE
How can I control data ttl in skywalking apm 9.5-9.7

### DISCUSSION THREAD

**Discussion by YifuYan** [apache/skywalking#11947] [Q&A]
I have upgraded skywalking version to my developing environment in kubernetes. But after running with a few days, old index doesnot deleted automatic. 
I have tried version 9.5.0 and 9.7.0. 
init mode and deployments settings:
![image](https://github.com/apache/skywalking/assets/38197640/a0cb194a-59cb-45ec-a595-7168747f3c8e)
ElasticSearch segment index:
![image](https://github.com/apache/skywalking/assets/38197640/670dcf7e-1951-4352-9bd2-1ff4de8d3bfb)

How can I fix this problem?
BTW: how does skywalking auto delete old index? use index ilm mechanism, or delete everyday by job?

**Comment by wu-sheng** [comment]
One of the OAPs should be selected as a leader, then it would delete the indices periodically.

**Reply by YifuYan** [reply]
Only one OAP server in the environment. Should I redeploy the OAP with cluster mode?

**Reply by wu-sheng** [reply]
If you are using standalone, it should be able to delete too.

**Reply by YifuYan** [reply]
When delete operation works, does it search all old indexes? or only search last day's index ?

**Reply by wu-sheng** [reply]
All indices matched the pattern.

**Reply by YifuYan** [reply]
Also, if change the TTL after a few days running, ex. 7 to 3, does it auto delete all old indexes?

**Reply by YifuYan** [reply]
Ok, I got it. But how can I fix my problem? Do I need to redeploy it?  When I update the version, I only changed the version in the k8s deployment definitions. Is this operation got this error?

**Reply by wu-sheng** [reply]
I don't know what is your side issue. It works on my side

**Reply by YifuYan** [reply]
Another relate issue, how to full delete all OAP's index (include index template). I don't want to redeploy es in my environment.
Have any docs for this delete operations?

**Reply by wu-sheng** [reply]
That is Elasticsearch command, not controlled by us.

**Reply by wu-sheng** [reply]
You could delete all skywalking prefix indices and templates.

**Reply by YifuYan** [reply]
yeah, I know that, does any index list or index tempalte list which is generated by oap server. I can remove all things in that list.
this will be helpful.

**Reply by wu-sheng** [reply]
It is, https://skywalking.apache.org/docs/main/latest/en/setup/backend/storages/elasticsearch/#elasticsearch

**Reply by YifuYan** [reply]
Thank you for your help. I will try to redeploy.

**Comment by YifuYan** [comment]
After redeployed oap server, there is many old index auto generated
![image](https://github.com/apache/skywalking/assets/38197640/c967cff7-095e-4a88-acf9-137261acdbc8)

In this environment, oap server fetch data from kafka, does the settings cause this issue?
![image](https://github.com/apache/skywalking/assets/38197640/465055be-2ede-4597-bdaf-8c7e9f4549d3)

**Reply by wu-sheng** [reply]
Maybe, I am not using Kafka(mentioned many times before). If your Kafka has old data, and OAP consumes from the beginning somehow, then yes. 

The indices are automatically recreated and driven by the data. TTL only runs once per day to remove expired ones.

**Reply by YifuYan** [reply]
When oap server does not strong enough ( no enough instance ), what can I use to replace kafka? May I have your suggest?

**Reply by wu-sheng** [reply]
If your OAP doesn't have enough resources, Kafka is making things worse. 

Quoted from FAQ

This question arises when users consider the circumstances where the OAP cluster may not be powerful enough or becomes offline. 
But the following issues must first be addressed:
1. Why do you think that the OAP is not powerful enough? Were it not powerful, the speed of data analysis wouldn't have caught up with the producers (or agents). Then what is the point of adding new deployment requirements?
1. Some may argue that the payload is sometimes higher than usual during peak times. But we must consider how much higher the payload really is.
1. If it is higher by less than 40%, how many resources would you use for the new MQ cluster? How about moving them to new OAP and ES nodes?
1. Say it is higher by 40% or more, such as by 70% to 200%. Then, it is likely that your MQ would use up more resources than it saves. 
Your MQ would support 2 to 3 times the payload using 10%-20% of the cost during usual times. Furthermore, in this case, 
if the payload/throughput are so high, how long would it take for the OAP cluster to catch up? The challenge here is that well before it catches up, the next peak times would have come.

With the analysis above in mind, why would you still want the traces to be 100%, given the resources they would cost? 
The preferred way to do this would be adding a better dynamic trace sampling mechanism at the backend. When throughput exceeds the threshold, gradually modify the active sampling rate from 100% to 10%, which means you could get the OAP and ES 3 times more powerful than usual, while ignoring the traces at peak times.

___
Source, https://skywalking.apache.org/docs/main/latest/en/faq/why_mq_not_involved/
